doc/reference/kernel/timing/clocks.rst

```diff
@@ -1,161 +1,242 @@
-
-Use services based on the hardware clock for time-based processing
-that requires higher precision than the system clock can provide,
-such as :ref:`busy_waiting` or fine-grained time measurements.
-
-.. note::
-    The high frequency of the hardware clock, combined with its 32-bit size,
-    means that counter rollover must be taken into account when taking
-    high-precision measurements over an extended period of time.
-
-Configuration
-*************
-
-Related configuration options:
-
-* :option:`CONFIG_SYS_CLOCK_TICKS_PER_SEC`
-
-API Reference
-*************
-
-.. doxygengroup:: clock_apis
-   :project: Zephyr
+Zephyr provides a robust and scalable timing framework to enable
+reporting and tracking of timed events from hardware timing sources of
+arbitrary precision.
+
+Time Units
+==========
+
+Kernel time is tracked in several units which are used for different
+purposes.
+
+Real time values, typically specified in milliseconds on microseconds,
+are the default presentation of time to application code.  They have
+the advantages of being universally portable and pervasively
+understood, though they may not match the precision of the underlying
+hardware perfectly.
+
+The kernel presents a "cycle" count via the ``k_cycle_get_32()`` API.
+The intent is that this counter represents the fastest cycle counter
+that the operating system is able to present to the user (for example,
+a CPU cycle counter) and that the read operation is very fast.  The
+expectation is that very sensitive application code might use this in
+a polling manner to achieve maximal precision.  The frequency of this
+counter is required to be steady over time, and is available from
+``sys_clock_hw_cycles_per_sec()`` (which on almost all platforms is a
+runtime constant that evaluates to
+CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC).
+
+For asynchronous timekeeping, the kernel defines a "ticks" concept.  A
+"tick" is the internal count in which the kernel does all its internal
+uptime and timeout bookeeping.  Interrupts are expected to be
+delivered on tick boundaries to the extent practical, and no
+fractional ticks are tracked.  The choice of tick rate is configurable
+via ``CONFIG_SYS_CLOCK_TICKS_PER_SEC``.  Defaults on most hardware
+platforms (ones that support setting arbitrary interrupt timeouts) are
+expected to be in the range of 10 kHz, with software emulation
+platforms and legacy drivers using a more traditional 100 Hz value.
+
+Conversion
+----------
+
+Zephyr provides an extensively enumerated conversion library with
+rounding control for all time units.  Any unit of "ms" (milliseconds),
+"us" (microseconds), "tick", or "cyc" can be converted to any other.
+Control of rounding is provided, and each conversion is available in
+"floor" (round down to nearest output unit), "ceil" (round up) and
+"near" (round to nearest).  Finally the output precision can be
+specified as either 32 or 64 bits.
+
+For example: ``k_ms_to_ticks_ceil32()`` will convert a millisecond
```

Comment:
I can't find this function.  Where is it defined?

Reply:
This section only applies to #17155 (and the names and API provided are in dispute).  I think once this gets through review it should be made part of that PR.